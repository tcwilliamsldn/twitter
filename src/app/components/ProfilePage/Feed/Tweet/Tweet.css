.tweet {
    display: flex;
    padding: 12px;
    border-bottom: 1px solid #f8f8f8;
}

.avatar {
    width: 60px;
    height: 60px;
    position: relative;
    border-radius: 50%;
    flex-shrink: 0;
    margin-right: 12px;
}

.user {
    font-size: 14px;
    margin-bottom: 12px;

    & > h3,
    & > h4 {
        margin-right: 4px;
    }

    & > h3 {
        font-weight: bold;
        display: inline;
        font-size: inherit;
    }

    & > h4 {
        color: #737373;
        display: inline;
        font-size: inherit;
        font-weight: normal;
    }

    & > time {
        color: #737373;

        &::before {
            content: '\00B7';
            margin-right: 4px;
        }
    }
}

.link {
    color: black;
    text-decoration: none;

    &:hover {
        text-decoration: underline;
    }
}
