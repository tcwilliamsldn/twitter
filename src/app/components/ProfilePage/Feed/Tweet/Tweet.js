import React from 'react';
import { Link } from 'react-router-dom';
import styles from './Tweet.css';
import enlargeImage from '../../../../utils/enlargeImage';

const Tweet = ({ text, created_at, id, user }) => {
    const { screen_name, name, profile_image_url } = user;

    const date = new Date(created_at).toLocaleString('en-gb', {
        month: 'short',
        day: 'numeric'
    });

    return (
        <div className={styles.tweet} key={id}>
            <img
                alt={`Profile picture of ${name}`}
                className={styles.avatar}
                src={enlargeImage(profile_image_url)}
            />
            <article>
                <div className={styles.user}>
                    <h3>{name}</h3>
                    <h4>@{screen_name}</h4>
                    <time dateTime={created_at}>{date}</time>
                </div>
                <section>{createMentionLinks(text, id)}</section>
            </article>
        </div>
    );
};

export function createMentionLinks(text) {
    const validUsername = /(@[a-zA-Z0-9_]+)/gi;
    const splitText = text.split(validUsername);

    const processedText = splitText.map(
        (section, index) =>
            section.startsWith('@') ? (
                <Link
                    className={styles.link}
                    key={`${section}${index}`}
                    to={`${section.substring(1)}`}
                >
                    {section}
                </Link>
            ) : (
                section
            )
    );

    return processedText;
}

export default Tweet;
