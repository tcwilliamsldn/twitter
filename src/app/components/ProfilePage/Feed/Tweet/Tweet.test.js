import Tweet, { createMentionLinks } from './Tweet';
import React from 'react';
import { shallow } from 'enzyme';

describe('Tweet', function() {
    it('renders a Tweet', function() {
        const props = {
            text: 'Hello World',
            created_at: 'Fri Jul 20 02:17:02 +0000 2018',
            id: '123',
            user: {
                screen_name: 'tom',
                name: 'Tom',
                profile_image_url: 'http://example.com/image.jpg'
            }
        };

        const component = shallow(<Tweet {...props} />);
        expect(component).toMatchSnapshot();
    });
});

describe('createMentionLinks', function() {
    it('turns mentions into Links, that link to the mentioned users profile', function() {
        const text = 'Hello @person what a great day';
        const processedText = createMentionLinks(text);
        expect(processedText).toMatchSnapshot();
    });

    it('ignores @s in the middle of a word', function() {
        const text = 'H@ters gonna H@te';
        const processedText = createMentionLinks(text);
        expect(processedText).toMatchSnapshot();
    });

    it('ignores @s in the middle of a word and still adds valid links', function() {
        const text = 'H@ters gonna H@te @Tom';
        const processedText = createMentionLinks(text);
        expect(processedText).toMatchSnapshot();
    });

    it('leaves non-username trailing letters out of the link', function() {
        const text = '@Person! @Person? @Person, how are you? RT @Person: Link';
        const processedText = createMentionLinks(text);
        expect(processedText).toMatchSnapshot();
    });
});
