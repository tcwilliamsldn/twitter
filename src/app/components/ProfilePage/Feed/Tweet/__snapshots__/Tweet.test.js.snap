// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet renders a Tweet 1`] = `
<div
  className="tweet"
  key="123"
>
  <img
    alt="Profile picture of Tom"
    className="avatar"
    src="http://example.com/image.jpg"
  />
  <article>
    <div
      className="user"
    >
      <h3>
        Tom
      </h3>
      <h4>
        @
        tom
      </h4>
      <time
        dateTime="Fri Jul 20 02:17:02 +0000 2018"
      >
        Jul 20
      </time>
    </div>
    <section>
      Hello World
    </section>
  </article>
</div>
`;

exports[`createMentionLinks ignores @s in the middle of a word 1`] = `
Array [
  "H",
  <Link
    className="link"
    replace={false}
    to="ters"
  >
    @ters
  </Link>,
  " gonna H",
  <Link
    className="link"
    replace={false}
    to="te"
  >
    @te
  </Link>,
  "",
]
`;

exports[`createMentionLinks ignores @s in the middle of a word and still adds valid links 1`] = `
Array [
  "H",
  <Link
    className="link"
    replace={false}
    to="ters"
  >
    @ters
  </Link>,
  " gonna H",
  <Link
    className="link"
    replace={false}
    to="te"
  >
    @te
  </Link>,
  " ",
  <Link
    className="link"
    replace={false}
    to="Tom"
  >
    @Tom
  </Link>,
  "",
]
`;

exports[`createMentionLinks leaves non-username trailing letters out of the link 1`] = `
Array [
  "",
  <Link
    className="link"
    replace={false}
    to="Person"
  >
    @Person
  </Link>,
  "! ",
  <Link
    className="link"
    replace={false}
    to="Person"
  >
    @Person
  </Link>,
  "? ",
  <Link
    className="link"
    replace={false}
    to="Person"
  >
    @Person
  </Link>,
  ", how are you? RT ",
  <Link
    className="link"
    replace={false}
    to="Person"
  >
    @Person
  </Link>,
  ": Link",
]
`;

exports[`createMentionLinks turns mentions into Links, that link to the mentioned users profile 1`] = `
Array [
  "Hello ",
  <Link
    className="link"
    replace={false}
    to="person"
  >
    @person
  </Link>,
  " what a great day",
]
`;
